### ISSUE TITLE
Remediation: Unicode Book titles mangled during Import

### ISSUE THREAD

**Issue by anandology** [internetarchive/openlibrary#135]
This issue is reported in the ol-tech mailing list.

```
I don't know how widespread this problem is, but I noticed that these
two records have messed up book titles, but if you click through to
the associated MARC records on IA, the titles get rendered correctly.

http://openlibrary.org/books/OL7155555M/The_M%C2%A9%C3%98alavik%C2%A9%C3%98agnimitra
http://openlibrary.org/books/OL7165183M/The_Vikramorva%C2%A9%C3%98s%C2%A9%C4%90iyam
```

**Comment by bfalling** [comment]
Should we consider re-importing? And is #149, which also references https://bugs.launchpad.net/openlibrary/+bug/598204, a dependency?

**Comment by hornc** [comment]
Some I have found recently: https://openlibrary.org/works/OL17670297W
https://openlibrary.org/works/OL17677126W
multiple works by this author:
https://openlibrary.org/authors/OL2450531A/Matthieu_Joseph_Bonaventure_Orfila

**Comment by LeadSongDog** [comment]
https://openlibrary.org/search?q=title%3A+%22©♭%22&mode=everything Still finds well over 17 million matches. This for "é", likely the most common accented letter. Edits like https://openlibrary.org/books/OL26303038M/Anatomie_générale_appliquée_à_la_physiologie_et_à_la_médecine?b=3&a=1&_compare=Compare&m=diff should not be necessarily manual.

**Comment by LeadSongDog** [comment]
@hornc re your May 8 comment, those works were created from editions created from importing
https://openlibrary.org/show-records/ia:b28044277_0001
and 
https://openlibrary.org/show-records/ia:b2202010x
Until they're fixed in the ia MARC records, there's no value in reimporting unless the importation gets them passed through normalization

**Comment by hornc** [comment]
@LeadSongDog interesting, the MARC display you link to shows the characters garbled, but if you click through to the XML representation https://ia800202.us.archive.org/34/items/b28044277_0001/b28044277_0001_marc.xml  the accented e's display correctly. There may be an issue with encoding types set incorrectly? I will pick this up shortly, the new openlibrary-client is now in a state where it can be used to make bulk data corrections.

**Comment by hornc** [comment]
@LeadSongDog I may have figured out how the mangling is happening, in this example marc xml
https://ia600208.us.archive.org/25/items/b2202010x/b2202010x_marc.xml

the a-grave of "Secours à donner" displays correctly in utf-8 encoding

a-grave is U+00E0, which in binary (pythonic notation) is `\xC3\xA0`

if those bytes were interpreted as MARC8 and "converted", `C3` becomes the copyright symbol, and 'A0' becomes a space, which is exactly what we see on the OL pages with "Secours ©  donner"

I now think these MARC records have utf-8 character encodings, but were imported to OL as if they were MARC8, which explains the mangling.

I did the MARC8 conversion manually from the tables found here https://memory.loc.gov/diglib/codetables/45.html I'll need to use yaz or something to test this out properly, but this will provide a good path to fixing the MARC errors programatically.

I know that there are other unicode mangling errors affecting Amazon imported records, but I think that is from incorrect conversion from Windows or ISO charsets

Thanks for your comment @LeadSongDog, in trying to figure out whether the MARC records were actually wrong or not I think I have stumbled upon the root cause of the issue!

**Comment by mekarpeles** [comment]
@hornc any updates on MARC mangling and/or if we resolved this issue?

**Comment by tfmorris** [comment]
The issue is definitely not resolved. When the import script is fixed, @bfalling suggest of reimporting will most likely be necessary.

From the point of view of triage, it would probably be useful to get an actual count. "Thousands" isn't a very big percentage of 25 million editions.

**Comment by cclauss** [comment]
Has this been resolved with our Python 3 changes or can someone provide steps-to-reproduce on Python 3?

**Comment by LeadSongDog** [comment]
Well https://openlibrary.org/books/OL12903648M/Etudes_Conomiques_De_L'Ocde certainly isn’t fixed, but perhaps we’re done with digging the hole...  
There were at least three problem classes:
1. Bad import of good data
2. Literal import of bad data
3. Bad data in place from old cases of 1 or 2 since remedied. 
The move to py3 will at most fix number 1.

**Comment by cclauss** [comment]
Steps-to-reproduce problem class 1?

**Comment by tfmorris** [comment]
The earlier examples are better than the most recent one which is an import from crappy Amazon data (which we should **not** be imported).
https://openlibrary.org/books/OL7165183M/The_Vikramorva%C2%A9%C3%98s%C2%A9%C4%90iyam
https://openlibrary.org/authors/OL4459814A/Heinrich_Schro_der
https://openlibrary.org/books/OL13956174M/Tonbandgera_te-Messpraxis
https://openlibrary.org/books/OL26280693M/Secours_%C2%A9_donner_aux_personnes_empoisonn%C2%A9%E2%99%ADes_ou_asphyxi%C2%A9%E2%99%ADes_suivis_des_moyens_propres_%C2%A9_reconna%C2%A9%CA%BEtre

If the bug has been fixed, reimporting the records should result in the correct encoding. Then the task simply becomes reimporting the millions of corrupted records.

The search which was claimed to return 17+ million records before: https://openlibrary.org/search?q=title%3A+%22%C2%A9%E2%99%AD%22&mode=everything
now returns 23.4M results, but I think that's actually a separate bug and it's just returning all works in the database.

**Comment by LeadSongDog** [comment]
@tfmorris As https://openlibrary.org/search?q=title%3A+%22+%22&mode=everything gets the same result, it seems that yes, it is a simple case of title search for an effectively-blank string.

**Comment by tfmorris** [comment]
I created #4223 for the search bug.

**Comment by hornc** [comment]
Update, running a current search
https://openlibrary.org/search?q=%C2%A9%E2%99%AD&mode=everything
with the new "Solr Editions Beta" checkbox  _disabled_ gives 1,798 results. (0 results with the beta feature enabled -- I do 't know what the expected difference should be).

* I believe all MARC record character encoding import issues are resolved, so there is just data clean up to perform.
* The number of these affected records seems to have decreased considerably over time, so that is good.

I'll try to come up with some clear queries to identify classes of MARC encoding caused character issues 
* Accents etc incorrectly moved from MARC encoding to UTF-8 like ©♭
* Spaces and dropped letters in place of an accented character

I think these issues are limited to work/edition titles and author names, but it'd pay to check whether locations or publishers are affected too.

I'm also not 100% sure a reimport will overwrite all the fields if they are already populated, unfortunately.

Garbled Amazon titles should be dealt with separately as the cause of mangling is more variable there.

I'll take a another look at MARC sourced mangled data remediation and see how much is left to close this after a decade...

**Comment by cdrini** [comment]
I'm finding about 21,000 works affected on open library: https://openlibrary.org/search?q=title_suggest%3A%C2%A9+AND+ia%3A*&mode=everything

**Comment by tfmorris** [comment]
In the 2 1/2 years since @hornc did his [analysis](https://github.com/internetarchive/openlibrary/wiki/Mangled-MARC), the counts have gone up about 10% to 
```
 1359 authors*
 9959 works
13694 editions
```
*the vast majority of the hits are on post-merge AKAs, not the primary name. MARCbot fixed a large number of author records in November 2025 e.g. https://openlibrary.org/authors/OL7397629A?_compare=Comparer&b=2&a=1&m=diff
also 

It looks like bad records are still being imported, with this one as recently as 26 Mar 2026
https://openlibrary.org/authors/OL16244558A?v=1

@hornc Perhaps it would make sense to include the signature bad character bigrams as a filter in the import pipeline to trigger an error or warning.

**Comment by mekarpeles** [comment]
**Status overview**

The import-side transcoding bug appears to be fixed (Python 3 migration resolved MARC-8 → UTF-8 mishandling), but bad records continue to arrive — @tfmorris noted in April 2026 that counts increased ~10% over 2.5 years to roughly **13,694 editions / 9,959 works / 1,359 authors**. The importer fix prevents new damage; the existing data still needs a cleanup pass.

**What's already done**
- Root cause documented: https://github.com/internetarchive/openlibrary/wiki/Mangled-MARC
- MARCbot corrected a large batch of author records in Nov 2025
- Solr edition beta shows 0 results for the `©♭` pattern — suggesting the indexed editions are partially remediated, but the underlying records are not

**Remaining work**

1. **Confirm importer is clean** — verify the current import pipeline doesn't still produce `©♭` artifacts for MARC-8 encoded sources (re: @tfmorris's April 2026 note that bad records are still arriving)
2. **Bulk remediation** — for the ~14k affected editions, re-import from their IA MARC source records or run a targeted data patch script. Reference: `openlibrary/catalog/marc/marc_binary.py` and `openlibrary/catalog/marc/parse.py` for the MARC parsing path
3. **Regression guard** — add a test with a known-mangled MARC-8 fixture to prevent recurrence

**Relevant files**
- `openlibrary/catalog/marc/marc_binary.py` — MARC-8 parsing
- `openlibrary/catalog/marc/parse.py` — field extraction
- `openlibrary/catalog/add_book/__init__.py` — import entry point

**Before starting:** confirm with @mekarpeles whether the approach is (a) re-import from source MARC, (b) a standalone data-patch script, or (c) both. The IA MARC records appear to have correct UTF-8 when accessed as XML — re-import via openlibrary-client may be feasible.

— Richy, via PAM (Open Library's Project AI Manager)
